### ISSUE TITLE
Captain.log - IPLD v1 spec

### ISSUE THREAD

**Issue by nicola** [ipld/specs#13]
## 31-6 Aug

Fellow explorers, I am jumping on this spaceship, taking control on the pilot cabin and claim myself captain. Thank you to all the people that are contributing and have contributed on IPLD, as well to everyone using and willing to use IPLD. In this repo we will "make IPLD great again" so that it can truly be bring interoperability, secure permanent links, authenticated data structure to the whole web (worldwide and interplanetary).

I will keep this issue updated weekly (or so), I will write about (1) what is happening, (2) where help is needed the most. Please **subscribe to this issue**, or **watch the repo** for updates.
#### What is happening
- [x] We are refining the spec and we have a list of issues to be solved and we need to reach on agreements:
  - #1 : Relative paths: yes or not?
    - yes: (pro) circular relations!, (cons) new errors (parent does not exist!)
    - no: (pro) simple design!, (cons) no circular relations across objects
  - #2 : Properties in links: can link have user properties beyond `/`?
    - yes: (pro) can have hidden properties, (cons) complex!
    - no: (pro) simple! can use this space for IPLD specific keywords, (cons) NOTHING!
  - #3 : Support links to have pointers that are `hash` + `/path/to/whatever`
  - #5 : Can I specify multiple links with different hash functions in the link object?
    - yes: (pro) hash can break, links still exist (cons) need to detect if links' references are equal, complex typing
    - no: (pro) clean link object (cons) no survival through hash break
  - #6 : CBOR should be upgradeable
  - #7 : Namespace for IPLD
  - #8 : Should IPLD object point to IPLD objects _only_?
  - #9 : How to handle mutable links
  - #11 : Errors to be handled
- New writing of the spec is happening: here is a draft #10 
- IPLD moved into its own repo (however it is still in progress https://github.com/ipfs/specs/issues/128 - help needed by @dignifiedquire, @jbenet, @diasdavid in moving all the things)
- New proposal: CID and multicodec-packed https://github.com/ipfs/specs/issues/130
#### Where is the help needed
- [x] We are having a call next monday to discuss the above https://github.com/ipfs/pm/issues/124
- There is a need to agree on most of the issues above
- Provide examples of your use cases and needs

This is all for now!
#### Thank-yous
- Thank you to @dignifiedquire and @RichardLitt for going over some of the previous PRs and issues
- Thanks to @diasdavid and @jbenet for the updates on CID & co
## 

cc @mildred, @stebalien, @dignifiedquire

**Comment by nicola** [comment]
## 7-14 Aug

I will write this log in advance so that I don't miss pieces!

#### tl;dr
- Issues that need **[more discussion](https://github.com/ipld/specs/issues?q=is%3Aissue+is%3Aopen+label%3Adiscussion)**
- Issues that we **[already agreed upon](https://github.com/ipld/specs/issues?q=is%3Aissue+is%3Aopen+label%3A%22needs+spec%22)**

#### What happened (done)
- We did a call on Monday and here are the [minutes](https://github.com/ipfs/pm/issues/124#issuecomment-238724365)
- Finalized agreement on:
  - 🎉 #6 : Spec refining: Do not bind the canonical representation to CBOR needs spec spec
  - 🎉 #5 : Spec refining: Multiple links in the link object needs spec spec
  - 🎉 #3 : Spec refining: Support of IPLD pointers as links needs spec spec
  - 🎉 #2 : Spec refining: Properties in links needs spec spec

#### What is happening (todo)
- [x] https://github.com/ipld/specs/pull/12 : Spec: Getting DRAFT of v1 IPLD done (at least the skeleton)
- [x] https://github.com/ipld/specs/issues/11 : Spec: Listing down errors to be handled
- [x] https://github.com/ipld/specs/issues/15 : Spec: Terminology prefixing everything with IPLD or with Merkle (bike-shedding post)
- [x] https://github.com/ipld/ipld.io/issues/1 : Website: Starting up IPLD.io

##### Need to find agreement from past week
- [ ] https://github.com/ipld/specs/issues/1 : Relative paths: yes or not? (in new comments @nicola argues on why it is important)
- https://github.com/ipld/specs/issues/8 : Forcing IPLD to only point to IPLD paths (apparently it is a no, but need more agreement)
- https://github.com/ipld/specs/issues/7 : Requiring prefixing (full CID and/or namespace) can this be implicit?
- https://github.com/ipld/specs/issues/9 : Mutable links on IPLD (although we kind of agreed that we want them, but we need a distinguisher and this could be the namespace, needs more agreement)
- https://github.com/ipfs/specs/issues/130 : CID (we seems to agree that it solves a lot of problems)

#### Where is the help needed
- We still have pending issues over here: [issues labelled spec](https://github.com/ipld/specs/issues?q=is%3Aissue+is%3Aopen+label%3Aspec)
- https://github.com/ipfs/notes/issues/152 : Bringing RDF on IPLD (or authenticating RDF)
- https://github.com/ipfs/notes/issues/154 : PubSub & corecursive IPFS: how to listen to events, maybe related to IPLD

#### Thank-yous
- Thank you to @dignifiedquire for helping with the tooling so that we now have `needs spec` label if we agree and we just need to write the spec
- Thank you to @em-ly @jbenet @dignifiedquire @diasdavid @EdwardSilverton @flyingzumwalt @stebalien for joining the call and participate to this week's activities

**Comment by nicola** [comment]
# Aug 15-22

It has been a calm week. A lot of conversations. We are planning a new call soon!

#### What happened (done)
- merged https://github.com/ipld/specs/pull/21 https://github.com/ipld/specs/pull/22 thanks to @tjgillies and @EliasGabrielsson 

#### What is happening (todo)
- [x] https://github.com/ipld/specs/issues/19 : IPNS+snapshot links discussion
- [x] https://github.com/ipld/specs/issues/18 : Conversation whether IPLD is too big to be the thin layer
- see the unchecked previous todos for the work TODO!

##### What is happening from past weeks
- [x] https://github.com/ipld/specs/pull/12 : Spec: Getting DRAFT of v1 IPLD done (at least the skeleton)
- [ ] https://github.com/ipld/ipld.io/issues/1 : Website: Starting up IPLD.io

##### Need to find agreement from past weeks

See previous post, but as a summary:
- https://github.com/ipld/specs/issues/15 : Spec: Terminology prefixing everything with IPLD or with Merkle (bike-shedding post)
-  https://github.com/ipld/specs/issues/11 : Spec: Listing down errors to be handled
-  https://github.com/ipld/specs/issues/1 : Relative paths: yes or not? (in new comments @nicola argues on why it is important)
- https://github.com/ipld/specs/issues/8 : Forcing IPLD to only point to IPLD paths (apparently it is a no, but need more agreement)
- https://github.com/ipld/specs/issues/7 : Requiring prefixing (full CID and/or namespace) can this be implicit?
- https://github.com/ipld/specs/issues/9 : Mutable links on IPLD (although we kind of agreed that we want them, but we need a distinguisher and this could be the namespace, needs more agreement)
- https://github.com/ipfs/specs/issues/130 : CID (we seems to agree that it solves a lot of problems)

#### What is going on in the community
- 👪  https://github.com/mediachain/mediachain/issues/158 : IPLD and Mediachain!
- 👪 https://github.com/NixOS/nix/issues/859 : IPLD/FS and NixOS

#### Where is the help needed
- We still have pending issues over here: [issues labelled spec](https://github.com/ipld/specs/issues?q=is%3Aissue+is%3Aopen+label%3Aspec)
- https://github.com/ipfs/notes/issues/152 : Bringing RDF on IPLD (or authenticating RDF)
- https://github.com/ipfs/notes/issues/154 : PubSub & corecursive IPFS: how to listen to events, maybe related to IPLD

#### Thank-yous
- @stebalien, @jbenet, @tjgillies, @harlantwood, @Ericson2314

**Comment by jbenet** [comment]
(Awesome captain log! Cc @whyrusleeping @diasdavid)

**Comment by nicola** [comment]
# The captain is off the ship until the 5th September

**Comment by nicola** [comment]
# 5-11 Sept

Academic term has started, devs are going back to code, this is a great time to start with all the energy that we collected during the summer.

---

### What happened (done)

Not much has happened I and other contributors have been out and silent

### What is happening (todo)
- see the unchecked previous todos for the work TODO!
- 📞  https://github.com/ipfs/pm/issues/181 : New call coming up 

#### What is happening from past weeks
- [ ] https://github.com/ipld/specs/pull/12 : Spec: Getting DRAFT of v1 IPLD done (at least the skeleton)
- [ ] https://github.com/ipld/ipld.io/issues/1 : Website: Starting up IPLD.io

#### Need to find agreement from past weeks

See previous unchecked issues

## 

### What is going on in the community
- 👪 https://github.com/ipfs/notes/issues/161 : Persistent balanced trees on IPFS

### Where is the help needed
- We still have pending issues over here: [issues labelled spec](https://github.com/ipld/specs/issues?q=is%3Aissue+is%3Aopen+label%3Aspec)
- https://github.com/ipfs/notes/issues/152 : Bringing RDF on IPLD (or authenticating RDF)
- https://github.com/ipfs/notes/issues/154 : PubSub & corecursive IPFS: how to listen to events, maybe related to IPLD

---

### Thank-yous

Thank you to no one in particular :)

**Comment by nicola** [comment]
# 11-18 Sept

A lot of conversations on relative links, how to have circular relations (on a data structure level and on an application level). This happened during the call https://github.com/ipfs/pm/issues/181, (notes are here: https://github.com/ipfs/pm/pull/196)

---

### What is happening this week
- [ ] #12 : (Finally) Writing the skeleton of the new IPLD v1 spec
- [ ] 📞 https://github.com/ipfs/pm/issues/197 : There will be a call squeezed in the IPFS (going to happen after libp2p)

### What happened (done)
- [x] #18 : Is this thin?
- [x] #17 : No more @link.

### To Agree
- [ ] #15 : Spec refining: Terminology IPLD vs Merkle
- [ ] #11 : Spec refining: listing down the possible errors to be handled
- [ ] #9 : Spec refining: Mutable links in IPLD
- [ ] #8 : Spec refining: IPLD pointers must point to IPLD objects only
- [ ] #7 : Spec refining: requiring namespace in IPLD pointers

### Agreed (needs spec writing)
- :tada: #1 : Spec refining: Relative paths in IPLD !!!!
- :tada: #20 : Spec refining: Merkle-paths to an array's index
- :tada: #6 : Spec refining: Do not bind the canonical representation to CBOR
- :tada: #5 : Spec refining: Multiple links in the link object
- :tada: #3 : Spec refining: Support of IPLD pointers as links
- :tada: #2 : Spec refining: Properties in links

### Others (proposals, question, ..)
- [ ] #19 : Idea for permanent mutable links

---

## Thank-yous
- @Stebalien for writing the minutes of the previous call (and the work on cycles/relative paths/co)

**Comment by nicola** [comment]
# 19-02 Oct

Now there are weekly meetings on IPLD (see [ipfs/pm](https://github.com/ipfs/pm/issues)) for the current sprint. A lot of progress on how IPLD code structure should look like and several conversations regarding relative paths (again!)

### What is happening this week
- [ ] Call in which we need an update from @diasdavid and @jbenet on CID and IPLD interface & new 🚲 🏚 !@
- [ ] Going over a skeleton for the new spec

### What happened (ongoing)
- [ ] #27 : Map the entire Ethereum State into IPFS with ipld
- [ ] #26 : Implementation: Make resolver into its own module
- [ ] #25 : Implementation: separate IPLD from CBOR
- [ ] #24 : Implementation: Tracking implementation effort for IPLD v1
- [ ] #23 : IPLD v2: Laying out Future plans

### What happened (done)
- [x] #28 : Fix bracket mismatch in example.

### To Agree
- [ ] #15 : Spec refining: Terminology IPLD vs Merkle
- [ ] #11 : Spec refining: listing down the possible errors to be handled
- [ ] #9 : Spec refining: Mutable links in IPLD
- [ ] #8 : Spec refining: IPLD pointers must point to IPLD objects only
- [ ] #7 : Spec refining: requiring namespace in IPLD pointers

### Agreed (needs spec writing)
- :tada: #20 : Spec refining: Merkle-paths to an array's index
- :tada: #6 : Spec refining: Do not bind the canonical representation to CBOR
- :tada: #5 : Spec refining: Multiple links in the link object
- :tada: #3 : Spec refining: Support of IPLD pointers as links
- :tada: #2 : Spec refining: Properties in links
- :tada: #1 : Spec refining: Relative paths in IPLD

### Others (proposals, question, ..)
- [ ] #19 : Idea for permanent mutable links

### Thanks you-s

@flyingzumwalt, @dignifiedquire, @RichardLitt

**Comment by daviddias** [comment]
@nicola update:

![image](https://cloud.githubusercontent.com/assets/1211152/19034617/6368898c-895c-11e6-9b03-786dbbaf2b6b.png)

See https://github.com/ipfs/js-ipfs-ipld/pull/60, we got the interface down :D 

>  #27 : Map the entire Ethereum State into IPFS with ipld

It is not going to be mapped, it is going to be resolved natively, thanks to "IPLD formats" :) 

> Implementation: separate IPLD from CBOR

We got this with the creation of the interface-ipld-format and the IPLD Formats identified by multicodecs https://github.com/multiformats/multicodec/pull/16
